### PR TITLE
CEP-18 

### PR DESCRIPTION
**Pull request by kubaplas**
Implementation of CEP-18 in Odra.
Named keys implementation is still missing.

## Summary by CodeRabbit

- **New Features**
  - Introduced new functionalities for managing named and dictionary values in contracts, enhancing interaction capabilities.
  - Deployed and tested CEP-18 compliant token contracts, providing full lifecycle management including token transfers and approvals.
- **Bug Fixes**
  - Added checks to prevent contract execution from a contract address, enhancing security.
- **Documentation**
  - Updated documentation related to contract execution and error handling.
- **Refactor**
  - Optimized contract interaction methods and refactored storage handling for efficiency and clarity.
- **Tests**
  - Added comprehensive test cases for token minting, burning, and transfer scenarios to ensure robustness and security.
- **Chores**
  - Updated dependencies and integrated new modules for enhanced functionality.
- **Style**
  - Improved code readability and maintenance by standardizing storage and error handling modules.

### CODE REVIEW

File: modules/src/cep18_token.rs [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+            self.env().revert(CannotTargetSelfUser);
+        }
+        self.raw_transfer(&caller, recipient, amount);
+    }
+
+    /// Transfers tokens from the owner to the recipient using the spender's allowance.
+    pub fn transfer_from(&mut self, owner: &Address, recipient: &Address, amount: &U256) {
+        let spender = self.env().caller();
+
+        if *owner == *recipient {
```
Comment (kpob) [inline modules/src/cep18_token.rs]: Compare refs
```suggestion
        if owner == recipient {
```

File: modules/src/cep18/utils.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+}
+/// Modality of the CEP-18 contract.
+#[odra::odra_type]
+pub enum Cep18Modality {
+    /// No modailities are set.
+    None = 0,
+    /// The contract can mint and burn tokens.
+    MintAndBurn = 1
+}
```
Comment (kpob) [inline modules/src/cep18/utils.rs]:
```suggestion
#[derive(Default)]
#[odra::odra_type]
pub enum Cep18Modality {
    /// No modalities are set.
    #[default]
    None = 0,
    /// The contract can mint and burn tokens.
    MintAndBurn = 1
}
```

File: modules/src/cep18_token.rs [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    TransferFrom, ChangeSecurity])]
+pub struct Cep18 {
+    decimals: SubModule<Cep18DecimalsStorage>,
+    symbol: SubModule<Cep18SymbolStorage>,
+    name: SubModule<Cep18NameStorage>,
+    total_supply: SubModule<Cep18TotalSupplyStorage>,
+    balances: SubModule<Cep18BalancesStorage>,
+    allowances: SubModule<Cep18AllowancesStorage>,
+    security_badges: Mapping<Address, SecurityBadge>,
+    modality: Var<u8>
```
Comment (kpob) [inline modules/src/cep18_token.rs]:
```suggestion
    modality: Var<Cep18Modality>
```

File: modules/src/cep18_token.rs [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    #[allow(clippy::too_many_arguments)]
+    pub fn init(
+        &mut self,
+        symbol: String,
+        name: String,
+        decimals: u8,
+        initial_supply: U256,
+        minter_list: Vec<Address>,
+        admin_list: Vec<Address>,
+        modality: Option<u8>
```
Comment (kpob) [inline modules/src/cep18_token.rs]: As `Cep18Modality` is odra type, and its `CLType` is `U8` I'd use something more meaningful than raw u8.
```suggestion
        modality: Option<Cep18Modality>
```

File: modules/src/cep18_token.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    /// Beware: do not remove the last Admin because that will lock out all admin functionality.
+    pub fn change_security(
+        &mut self,
+        admin_list: Vec<Address>,
+        minter_list: Vec<Address>,
+        none_list: Vec<Address>
+    ) {
+        // if mint burn is disabled, revert
+        if self.modality.get_or_default()
+            != <Cep18Modality as Into<u8>>::into(Cep18Modality::MintAndBurn)
```
Comment (kpob) [inline modules/src/cep18_token.rs]:
```suggestion
        if self.modality.get_or_default()  != Cep18Modality::MintAndBurn
```

File: modules/src/cep18_token.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        }
+
+        let allowance = self.allowance(owner, &spender);
+
+        self.allowances.set(
+            owner,
+            recipient,
+            allowance
+                .checked_sub(*amount)
+                .unwrap_or_revert_with(&self.env(), Error::InsufficientAllowance)
```
Comment (kpob) [inline modules/src/cep18_token.rs:233]: Sometimes you use `Error::`, sometimes not.

File: modules/src/lib.rs [hunk 7]
```diff
@@ -5,6 +5,8 @@
 extern crate alloc;
 
 pub mod access;
+pub mod cep18;
```
Comment (kpob) [inline modules/src/lib.rs:8]:
```suggestion
mod cep18;
```

File: modules/Cargo.toml [hunk 8]
```diff
@@ -12,6 +12,8 @@ categories = ["wasm"]
 [dependencies]
 odra = { path = "../odra", version = "0.9.1", default-features = false }
 hex = { version = "0.4.3", default-features = false }
+base64 = { version = "0.22.0", default-features = false, features = ["alloc"] }
+log = "0.4.20"
```
Comment (kpob) [inline modules/Cargo.toml]: remove

File: core/src/contract_env.rs [hunk 9]
```diff
@@ -175,18 +230,14 @@ impl ContractEnv {
     /// # Returns
     ///
     /// The hash value as a 32-byte array.
-    pub fn hash<T: ToBytes>(&self, value: T) -> [u8; 32] {
-        let bytes = value
-            .to_bytes()
-            .map_err(ExecutionError::from)
-            .unwrap_or_revert(self);
-        self.backend.borrow().hash(&bytes)
+    pub fn hash<T: AsRef<[u8]>>(&self, value: T) -> [u8; BLAKE2B_DIGEST_LENGTH] {
```
Comment (kpob) [inline core/src/contract_env.rs:230]: Why?

File: core/src/contract_env.rs [hunk 10]
```diff
@@ -1,12 +1,16 @@
+use casper_types::bytesrepr::deserialize_from_slice;
```
Comment (zie1ony) [inline core/src/contract_env.rs]: Unify imports of `casper_types`.

File: core/src/contract_env.rs [hunk 11]
```diff
@@ -240,24 +291,12 @@ impl ExecutionEnv {
...
             let bytes = self.env.backend.borrow().get_opt_named_arg_bytes(name);
-            let result = bytes.map(|bytes| deserialize_bytes(bytes, &self.env));
+            let result =
+                bytes.map(|bytes| deserialize_from_slice(bytes).unwrap_or_revert(&self.env));
             T::unwrap(result, &self.env)
         }
     }
 }
-
-fn deserialize_bytes<T: FromBytes>(bytes: Bytes, env: &ContractEnv) -> T {
```
Comment (zie1ony) [inline core/src/contract_env.rs:252]: Good catch with removing this fn.

File: core/src/contract_env.rs [hunk 12]
```diff
@@ -87,6 +91,57 @@ impl ContractEnv {
...
+    ) -> Option<T> {
+        let dictionary_name = dictionary_name.as_ref();
+        let key = key.as_ref();
+        let bytes = self
+            .backend
+            .borrow()
+            .get_dictionary_value(dictionary_name, key);
+        bytes.map(|b| {
+            deserialize_from_slice(b)
+                .map_err(|_| Formatting)
```
Comment (zie1ony) [inline core/src/contract_env.rs:120]: Why are you shadow the real error with `Formatting`?

File: core/src/contract_env.rs [hunk 13]
```diff
@@ -87,6 +91,57 @@ impl ContractEnv {
...
+    pub fn set_dictionary_value<T: CLTyped + ToBytes, U: AsRef<str>, V: AsRef<str>>(
+        &self,
+        dictionary_name: U,
+        key: V,
+        value: T
+    ) {
+        let dictionary_name = dictionary_name.as_ref();
+        let key = key.as_ref();
+        let cl_value = CLValue::from_t(value)
+            .map_err(|_| Formatting)
```
Comment (zie1ony) [inline core/src/contract_env.rs:135]: same here

File: core/src/error.rs [hunk 14]
```diff
@@ -240,3 +245,29 @@ pub enum EventError {
...
+    fn from(error: BytesReprError) -> Self {
+        match error {
+            BytesReprError::EarlyEndOfStream => ExecutionError::EarlyEndOfStream.into(),
+            BytesReprError::Formatting => ExecutionError::Formatting.into(),
+            BytesReprError::LeftOverBytes => ExecutionError::LeftOverBytes.into(),
+            BytesReprError::OutOfMemory => ExecutionError::OutOfMemory.into(),
+            BytesReprError::NotRepresentable => ExecutionError::NotRepresentable.into(),
+            BytesReprError::ExceededRecursionDepth => ExecutionError::ExceededRecursionDepth.into(),
+            _ => ExecutionError::Formatting.into()
+        }
```
Comment (zie1ony) [inline core/src/error.rs:271]: You can have only one `into`
```suggestion
        }.into()
```

File: examples/Cargo.toml [hunk 15]
```diff
@@ -8,6 +8,7 @@ odra = { path = "../odra", default-features = false }
 odra-modules = { path = "../modules", default-features = false }
 sha3 = { version = "0.10.6", default-features = false }
 odra-casper-livenet-env = { path = "../odra-casper/livenet-env", optional = true }
+base64 = "0.22.0"
```
Comment (zie1ony) [inline examples/Cargo.toml]: Looks like it's not needed.

File: modules/src/cep18/cep18_client_contract.rs [hunk 16]
```diff
@@ -0,0 +1,61 @@
+use crate::cep18_token::Cep18ContractRef;
```
Comment (zie1ony) [inline modules/src/cep18/cep18_client_contract.rs:1]: Is this really needed? We have the whole testing framework, that allows to test those functions in a cleaner way in tests.

File: modules/src/cep18/mod.rs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+//! # Cep18 Module
+//! A module that allows implementation of CEP-18 token contracts with Odra framework.
```
Comment (zie1ony) [inline modules/src/cep18/mod.rs]:
```suggestion
//! # CEP-18 Token implemented in Odra.
```

File: modules/src/cep18/storage.rs [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+use odra::prelude::*;
+use odra::{Address, UnwrapOrRevert};
+
+use crate::cep18::errors::Error::InvalidState;
+
+use base64::prelude::*;
+
+const ALLOWANCES_KEY: &str = "allowances";
+const BALANCES_KEY: &str = "balances";
+
```
Comment (zie1ony) [inline modules/src/cep18/storage.rs]:
```suggestion
```

File: modules/src/cep18_token.rs [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+            allowance,
+            inc_by: *inc_by
+        });
+    }
+
+    /// Transfers tokens from the caller to the recipient.
+    pub fn transfer(&mut self, recipient: &Address, amount: &U256) {
+        let caller = self.env().caller();
+        if caller == *recipient {
+            self.env().revert(CannotTargetSelfUser);
```
Comment (zie1ony) [inline modules/src/cep18_token.rs]: No `if_zero()` check?

File: modules/src/cep18_token.rs [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+            amount: *amount
+        });
+
+        self.raw_transfer(owner, recipient, amount);
+    }
+
+    /// Mints new tokens and assigns them to the given address.
+    pub fn mint(&mut self, owner: &Address, amount: &U256) {
+        // check if mint_burn is enabled
+        if self.modality.get_or_default() == 0 {
```
Comment (zie1ony) [inline modules/src/cep18_token.rs]: You could encapsulate this logic inside `modality` like
```
self.modality.assert_minting_enabled();
```

File: modules/src/cep18_token.rs [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+            self.env().revert(MintBurnDisabled);
+        }
+
+        // check if the caller has the minter badge
+        let security_badge = self
```
Comment (zie1ony) [inline modules/src/cep18_token.rs:250]: Same here
```
self.security_badges.assert_caller_can_mint()
```

File: modules/src/cep18_token.rs [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+            self.env().revert(InvalidBurnTarget);
+        }
+
+        if self.balance_of(owner) < *amount {
+            self.env().revert(Error::InsufficientBalance);
+        }
+        let total_supply = self.total_supply.get();
+        let balance = self.balance_of(owner);
+
+        self.total_supply.set(
```
Comment (zie1ony) [inline modules/src/cep18_token.rs]: You could hide it in `self.total_supply.decrease(amount)` or in `self.decrease_total_supply(amount)`.

File: modules/src/cep18_token.rs [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+        };
+        setup_with_args(&env, init_args)
+    }
+
+    pub fn setup_with_args(env: &HostEnv, args: Cep18InitArgs) -> Cep18HostRef {
+        Cep18HostRef::deploy(env, args)
+    }
+
+    pub fn invert_address(address: Address) -> Address {
+        match address {
```
Comment (zie1ony) [inline modules/src/cep18_token.rs:362]: Why is this mambo jumbo needed?

File: odra-macros/src/ast/schema/custom_item.rs [hunk 24]
```diff
@@ -57,15 +57,15 @@ impl ToTokens for SchemaCustomTypeItem {
                         .chain(odra::prelude::vec![Some(#custom_item)])
                         .chain(odra::prelude::vec![#(Some(#enum_sub_types)),*])
                         #(#sub_types)*
-                        .collect::<Vec<_>>()
+                        .collect::<odra::prelude::Vec<_>>()
```
Comment (zie1ony) [inline odra-macros/src/ast/schema/custom_item.rs:60]: 👍🏻

File: odra-vm/src/vm/odra_vm.rs [hunk 25]
```diff
@@ -159,17 +162,39 @@ impl OdraVm {
         }
     }
 
+    /// Sets the value of the named key.
+    pub fn set_named_key(&self, name: &str, value: CLValue) {
+        // create a key by prepending the name with a prefix
+        let key = format!("{}_{}", NAMED_KEY_PREFIX, name);
```
Comment (zie1ony) [inline odra-vm/src/vm/odra_vm.rs]: Extract to another function, so the key creation is defined only once.

**Review by zie1ony — APPROVED** [review]
Well done :)

File: modules/src/cep18/utils.rs (hunk 2 above)
Comment (kubaplas) [reply modules/src/cep18/utils.rs]: Didn't know that you can do that...

File: modules/src/lib.rs (hunk 7 above)
Comment (kubaplas) [reply modules/src/lib.rs:8]: Why?

File: modules/src/cep18_token.rs (hunk 23 above)
Comment (kubaplas) [reply modules/src/cep18_token.rs:362]: It is a copypasta of test  from original CEP-18

File: modules/src/cep18/cep18_client_contract.rs (hunk 16 above)
Comment (kubaplas) [reply modules/src/cep18/cep18_client_contract.rs:1]: It tests if the correct caller was used and the difference between a contract and account addresess. Same as above - it is a recreation of original test.

File: core/src/contract_env.rs (hunk 12 above)
Comment (kubaplas) [reply core/src/contract_env.rs:120]: #417 will fix that

File: modules/src/cep18_token.rs (hunk 21 above)
Comment (kubaplas) [reply modules/src/cep18_token.rs:250]: But then I would need to pass env to it. Poor stack!

Comment (zie1ony) [reply modules/src/cep18_token.rs:250]: `security_badges` is a `SubModule`, it already has an env.

**Review by zie1ony — APPROVED** [review]

File: modules/src/lib.rs (hunk 7 above)
Comment (kpob) [reply modules/src/lib.rs:8]: Why not? I see no reason to make it public.

**Review by kpob — APPROVED** [review]

**Pull request merged**